From 3ea590727a2a33ad7aa940343b1bf55da8a4b54c Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Tue, 1 Aug 2017 07:07:08 -0400
Subject: [PATCH] Memory leaks in WriteMSLImage

echo "" >output.msl
magick convert Memory-Leak-19_output_msl_1501504023.36 output.msl

leak memory

bug-debian: https://bugs.debian.org/870525
bug: https://github.com/ImageMagick/ImageMagick/issues/636
origin: https://github.com/ImageMagick/ImageMagick/commit/841f7b27dc88c685c61252d59b7e20e94c982456
---
 coders/msl.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/coders/msl.c b/coders/msl.c
index 7efe3302e..a7ec16e13 100644
--- a/coders/msl.c
+++ b/coders/msl.c
@@ -7840,12 +7840,25 @@ static MagickBooleanType ProcessMSLScript(const ImageInfo *image_info,
   }
   if (msl_info.exception->severity == UndefinedException)
     (void) xmlParseChunk(msl_info.parser," ",1,MagickTrue);
+  /*
+    Free resources.
+  */
   xmlFreeParserCtxt(msl_info.parser);
   (void) LogMagickEvent(CoderEvent,GetMagickModule(),"end SAX");
   msl_info.group_info=(MSLGroupInfo *) RelinquishMagickMemory(
     msl_info.group_info);
   if (*image == (Image *) NULL)
     *image=(*msl_info.image);
+  *msl_info.image_info=DestroyImageInfo(*msl_info.image_info);
+  msl_info.image_info=(ImageInfo **) RelinquishMagickMemory(
+    msl_info.image_info);
+  *msl_info.draw_info=DestroyDrawInfo(*msl_info.draw_info);
+  msl_info.draw_info=(DrawInfo **) RelinquishMagickMemory(msl_info.draw_info);
+  msl_info.image=(Image **) RelinquishMagickMemory(msl_info.image);
+  *msl_info.attributes=DestroyImage(*msl_info.attributes);
+  msl_info.attributes=(Image **) RelinquishMagickMemory(msl_info.attributes);
+  msl_info.group_info=(MSLGroupInfo *) RelinquishMagickMemory(
+    msl_info.group_info);
   if ((*msl_info.image)->exception.severity != UndefinedException)
     return(MagickFalse);
   return(MagickTrue);
